File: Python/OneLang/One/ErrorManager.py
from OneLangStdLib import *
from enum import Enum
import OneLang.One.Ast.Types as types
import OneLang.One.AstTransformer as astTrans
import OneLang.One.Ast.Statements as stats
import OneLang.Utils.TSOverviewGenerator as tSOvervGen
import OneLang.One.Ast.Expressions as exprs

class LOG_TYPE(Enum):
    INFO = 1
    WARNING = 2
    ERROR = 3

class CompilationError:
    def __init__(self, msg, is_warning, transformer_name, node):
        self.msg = msg
        self.is_warning = is_warning
        self.transformer_name = transformer_name
        self.node = node

class ErrorManager:
    def __init__(self):
        self.transformer = None
        self.current_node = None
        self.errors = []
        self.last_context_info = None
    
    def get_location(self):
        t = self.transformer
        
        par = self.current_node
        while isinstance(par, exprs.Expression):
            par = par.parent_node
        
        location = None
        if isinstance(par, types.Field):
            location = f'''{par.parent_interface.parent_file.source_path.path} -> {par.parent_interface.name}::{par.name} (field)'''
        elif isinstance(par, types.Property):
            location = f'''{par.parent_class.parent_file.source_path.path} -> {par.parent_class.name}::{par.name} (property)'''
        elif isinstance(par, types.Method):
            location = f'''{par.parent_interface.parent_file.source_path.path} -> {par.parent_interface.name}::{par.name} (method)'''
        elif isinstance(par, types.Constructor):
            location = f'''{par.parent_class.parent_file.source_path.path} -> {par.parent_class.name}::constructor'''
        elif par == None:
            pass
        elif isinstance(par, stats.Statement):
            pass
        else:
            pass
        
        if location == None and t != None and t.current_file != None:
            location = f'''{t.current_file.source_path.path}'''
            if t.current_interface != None:
                location += f''' -> {t.current_interface.name}'''
                if isinstance(t.current_method, types.Method):
                    location += f'''::{t.current_method.name}'''
                elif isinstance(t.current_method, types.Constructor):
                    location += f'''::constructor'''
                elif isinstance(t.current_method, types.Lambda):
                    location += f'''::<lambda>'''
                elif t.current_method == None:
                    pass
                else:
                    pass
        
        return location
    
    def get_current_node_repr(self):
        return tSOvervGen.TSOverviewGenerator.preview.node_repr(self.current_node)
    
    def get_current_statement_repr(self):
        return "<null>" if self.transformer.current_statement == None else tSOvervGen.TSOverviewGenerator.preview.stmt(self.transformer.current_statement)
    
    def reset_context(self, transformer = None):
        self.transformer = transformer
    
    def log(self, type, msg):
        t = self.transformer
        text = (f'''[{t.name}] ''' if t != None else "") + msg
        
        if self.current_node != None:
            text += f'''\n  Node: {self.get_current_node_repr()}'''
        
        location = self.get_location()
        if location != None:
            text += f'''\n  Location: {location}'''
        
        if t != None and t.current_statement != None:
            text += f'''\n  Statement: {self.get_current_statement_repr()}'''
        
        if self.last_context_info != None:
            text += f'''\n  Context: {self.last_context_info}'''
        
        if type == LOG_TYPE.INFO:
            console.log(text)
        elif type == LOG_TYPE.WARNING:
            console.error(f'''[WARNING] {text}\n''')
        elif type == LOG_TYPE.ERROR:
            console.error(f'''{text}\n''')
        else:
            pass
        
        if type == LOG_TYPE.ERROR or type == LOG_TYPE.WARNING:
            self.errors.append(CompilationError(msg, type == LOG_TYPE.WARNING, t.name if t != None else None, self.current_node))
    
    def info(self, msg):
        self.log(LOG_TYPE.INFO, msg)
    
    def warn(self, msg):
        self.log(LOG_TYPE.WARNING, msg)
    
    def throw(self, msg):
        self.log(LOG_TYPE.ERROR, msg)

File: Java/src/main/java/OneLang/One/ErrorManager/ErrorManager.java
package OneLang.One.ErrorManager;

import OneLang.One.Ast.Types.SourceFile;
import OneLang.One.Ast.Types.IInterface;
import OneLang.One.Ast.Types.IMethodBase;
import OneLang.One.Ast.Types.Method;
import OneLang.One.Ast.Types.IAstNode;
import OneLang.One.Ast.Types.Field;
import OneLang.One.Ast.Types.Property;
import OneLang.One.Ast.Types.Constructor;
import OneLang.One.Ast.Types.Lambda;
import OneLang.One.AstTransformer.AstTransformer;
import OneLang.One.Ast.Statements.Statement;
import OneLang.Utils.TSOverviewGenerator.TSOverviewGenerator;
import OneLang.One.Ast.Expressions.Expression;

import OneLang.One.AstTransformer.AstTransformer;
import OneLang.One.Ast.Types.IAstNode;
import java.util.List;
import OneLang.One.ErrorManager.CompilationError;
import OneLang.One.Ast.Expressions.Expression;
import OneLang.One.Ast.Types.Field;
import OneLang.One.Ast.Types.Property;
import OneLang.One.Ast.Types.Method;
import OneLang.One.Ast.Types.Constructor;
import OneLang.One.Ast.Statements.Statement;
import OneLang.One.Ast.Types.Lambda;
import java.util.ArrayList;
import OneStd.console;

public class ErrorManager {
    public AstTransformer transformer;
    public IAstNode currentNode;
    public List<CompilationError> errors;
    public String lastContextInfo;
    
    public String getLocation() {
        var t = this.transformer;
        
        var par = this.currentNode;
        while (par instanceof Expression)
            par = ((Expression)par).parentNode;
        
        String location = null;
        if (par instanceof Field)
            location = ((Field)par).parentInterface.getParentFile().sourcePath.path + " -> " + ((Field)par).parentInterface.getName() + "::" + ((Field)par).getName() + " (field)";
        else if (par instanceof Property)
            location = ((Property)par).parentClass.getParentFile().sourcePath.path + " -> " + ((Property)par).parentClass.getName() + "::" + ((Property)par).getName() + " (property)";
        else if (par instanceof Method)
            location = ((Method)par).parentInterface.getParentFile().sourcePath.path + " -> " + ((Method)par).parentInterface.getName() + "::" + ((Method)par).name + " (method)";
        else if (par instanceof Constructor)
            location = ((Constructor)par).parentClass.getParentFile().sourcePath.path + " -> " + ((Constructor)par).parentClass.getName() + "::constructor";
        else if (par == null) { }
        else if (par instanceof Statement) { }
        else { }
        
        if (location == null && t != null && t.currentFile != null) {
            location = t.currentFile.sourcePath.path;
            if (t.currentInterface != null) {
                location += " -> " + t.currentInterface.getName();
                if (t.currentMethod instanceof Method)
                    location += "::" + ((Method)t.currentMethod).name;
                else if (t.currentMethod instanceof Constructor)
                    location += "::constructor";
                else if (t.currentMethod instanceof Lambda)
                    location += "::<lambda>";
                else if (t.currentMethod == null) { }
                else { }
            }
        }
        
        return location;
    }
    
    public String getCurrentNodeRepr() {
        return TSOverviewGenerator.preview.nodeRepr(this.currentNode);
    }
    
    public String getCurrentStatementRepr() {
        return this.transformer.currentStatement == null ? "<null>" : TSOverviewGenerator.preview.stmt(this.transformer.currentStatement);
    }
    
    public ErrorManager()
    {
        this.transformer = null;
        this.currentNode = null;
        this.errors = new ArrayList<CompilationError>();
    }
    
    public void resetContext(AstTransformer transformer) {
        this.transformer = transformer;
    }
    
    public void log(LogType type, String msg) {
        var t = this.transformer;
        var text = (t != null ? "[" + t.getName() + "] " : "") + msg;
        
        if (this.currentNode != null)
            text += "\n  Node: " + this.getCurrentNodeRepr();
        
        var location = this.getLocation();
        if (location != null)
            text += "\n  Location: " + location;
        
        if (t != null && t.currentStatement != null)
            text += "\n  Statement: " + this.getCurrentStatementRepr();
        
        if (this.lastContextInfo != null)
            text += "\n  Context: " + this.lastContextInfo;
        
        if (type == LogType.Info)
            console.log(text);
        else if (type == LogType.Warning)
            console.error("[WARNING] " + text + "\n");
        else if (type == LogType.Error)
            console.error(text + "\n");
        else { }
        
        if (type == LogType.Error || type == LogType.Warning)
            this.errors.add(new CompilationError(msg, type == LogType.Warning, t != null ? t.getName() : null, this.currentNode));
    }
    
    public void info(String msg) {
        this.log(LogType.Info, msg);
    }
    
    public void warn(String msg) {
        this.log(LogType.Warning, msg);
    }
    
    public void throw_(String msg) {
        this.log(LogType.Error, msg);
    }
}